### DISCUSSION TITLE
Asset Checks on DBT Assets

### DISCUSSION THREAD

**Discussion by derHeinzer** [dagster-io/dagster#18558] [Q&A] [answered]
Hi all!

I would like to use Asset Checks on DBT assets. In trying that, I encountered, that the only official way seems to be either:

1. Use dbt generic tests on single columns.
2. Define asset checks that execute in an own op

I dont like both options, because
a) I want to use more specific tests than dbt generic ones.
b) I want to define multiple checks that all query the same or similar data from an bigquery dbt asset. Querying multiple times results in larger execution time and bq costs.

I was able to resolve this issue in quite a hacky way:

```python
import json
from random import randint
from unittest.mock import patch
from dagster import AssetCheckSpec, AssetCheckResult, AssetCheckSeverity, AssetKey, BackfillPolicy, DailyPartitionsDefinition, AssetExecutionContext
from dagster_dbt import dbt_assets, DbtCliResource
from dagster_dbt.asset_decorator import get_dbt_multi_asset_args as original_fn


def patch_fn(*args, **kwargs):
    deps, outs, internal_asset_deps, check_specs = original_fn(*args, **kwargs)
    asset_key = AssetKey(['dbt_bq', 'pdh', 'panel'])
    check_specs.extend(
        [
            AssetCheckSpec(name='warning_passes', asset=asset_key),
            AssetCheckSpec(name='warning_fails', asset=asset_key),
            AssetCheckSpec(name='error_passes', asset=asset_key),
            AssetCheckSpec(name='error_fails', asset=asset_key),
            AssetCheckSpec(name='tobis_check', asset=asset_key),
        ]
    )
    return deps, outs, internal_asset_deps, check_specs


with patch("dagster_dbt.asset_decorator.get_dbt_multi_asset_args", side_effect=patch_fn, wraps=original_fn):
    @dbt_assets(
        manifest=dbt_bq_manifest_path,
        dagster_dbt_translator=bq_translator,
        select="dbt_data_pipe_bq.pdh.tables.panel",
        partitions_def=DailyPartitionsDefinition(
            start_date="2023-01-01",
            timezone="Europe/Berlin",
        ),
        backfill_policy=BackfillPolicy.multi_run(
            max_partitions_per_run=25
        ),
    )
    def dbt_pdh_panel_asset(
            context: AssetExecutionContext, dbt_bq: DbtCliResource, bq_client: BqClient,
    ):
        start, end = context.partition_time_window
        dbt_vars = {
            "min_date": str(start.date()),
            "max_date": str(end.date()),
        }
        dbt_cli_args = [
                           "build",
                           "--vars",
                           json.dumps(dbt_vars)
                       ] 

        yield from dbt_bq.cli(dbt_cli_args, context=context).stream()
        
        yield AssetCheckResult(passed=True, check_name='warning_passes', severity=AssetCheckSeverity.WARN, metadata={'foo': 0, 'bar': 0})
        yield AssetCheckResult(passed=False, check_name='warning_fails', severity=AssetCheckSeverity.WARN, metadata={'foo': 2, 'bar': 13})
        yield AssetCheckResult(passed=True, check_name='error_passes', severity=AssetCheckSeverity.ERROR)
        yield AssetCheckResult(passed=False, check_name='error_fails', severity=AssetCheckSeverity.ERROR)
        tobis_check_result = randint(0, 2) > 0
        yield AssetCheckResult(passed=tobis_check_result, check_name='tobis_check', severity=AssetCheckSeverity.ERROR)
```

In a real world szenario, I would load data from the bq dbt model "dbt_data_pipe_bq.pdh.tables.panel" on which I would execute the Checks on. 

However this solution feels a little hacky and I would like to know, if there are better ways to achieve this? Or maybe if there are plans to improve the asset checks to be set in the dbt_assets decorator or via the translator object in near future?

Thanks for any piece of advise!

**Comment by johannkm** [comment, marked as answer]
Hi @derHeinzer, singular tests are on our radar. I filed a tracking issue here: https://github.com/dagster-io/dagster/issues/18595
The difficulty is when singular tests depend on multiple models. We plan to improve support for that with 1.6.

**Reply by derHeinzer** [reply]
Thanks for your reply and filing that ticket @johannkm!

I see your point with multi-model dependencies. What about handling that in the Translator class? In case of multidependencies, that are not handled explicitly, the tests (checks) are being skipped. Maybe we can add a warning in that case. I dont like the configuration in dbt's metadata, since the asset checks are purely on the dagster site. Dbt let alone the singular tests will be executable regardless on how many models they depend.

Additionally singular tests do not cover my szenario described above. Executing several dbt tests on a huge model might be more expensive then executing multiple asset checks in dagster. In dagster you might load the required data exactly once and execute as many checks on the loaded data as you like. Hence I would propose to introduce a way to add arbitrary asset check specifications to dbt assets via the decorator or the translator object. The so added AssetChecks are no dbt checks at all but do need to be yielded manually like in my example code above.
